---
title: "Symbiodiniaceae ITS2 intragenomic variants from cultures: Bioinformatic pipeline and sequence similarity networks"
author: "MR Nitschke"
date: "16/12/2019"
---
Following the dada2 workflow here https://benjjneb.github.io/dada2/tutorial.html, and here https://benjjneb.github.io/dada2/ITS_workflow.html

# Libraries

```{r}
library(dada2)
library(tidyverse)
theme_set(theme_bw())
library(ShortRead)
library(Biostrings)
library(phyloseq)
source("../help.R")
```

# Staging. Cutadapt for demultiplexing and barcode trimming

Run cutadapt on each MiSeq run.

```{}
# Complete!
cutadapt -g file:FTP110_barcodes_fwd.fasta -G file:FTP110_barcodes_rev.fasta -o demuxtrim-{name1}-{name2}.R1.fastq.gz -p demuxtrim-{name1}-{name2}.R2.fastq.gz MSRun156-FTP110_S1_L001_R1_001.fastq.gz MSRun156-FTP110_S1_L001_R2_001.fastq.gz

# Complete!
cutadapt -g file:FTP126_barcodes_fwd.fasta -G file:FTP126_barcodes_rev.fasta -o demuxtrim-{name1}-{name2}.R1.fastq.gz -p demuxtrim-{name1}-{name2}.R2.fastq.gz MSRun173-FTP126_S1_L001_R1_001.fastq.gz MSRun173-FTP126_S1_L001_R2_001.fastq.gz

# Complete!
cutadapt -g file:FTP127_barcodes_fwd.fasta -G file:FTP127_barcodes_rev.fasta -o demuxtrim-{name1}-{name2}.R1.fastq.gz -p demuxtrim-{name1}-{name2}.R2.fastq.gz MSRun174-FTP127_S1_L001_R1_001.fastq.gz MSRun174-FTP127_S1_L001_R2_001.fastq.gz

# Complete!
cutadapt -g file:FTP132_barcodes_fwd.fasta -G file:FTP132_barcodes_rev.fasta -o demuxtrim-{name1}-{name2}.R1.fastq.gz -p demuxtrim-{name1}-{name2}.R2.fastq.gz MSRun179-FTP132_S1_L001_R1_001.fastq.gz MSRun179-FTP132_S1_L001_R2_001.fastq.gz
```

# 1. DADA2

## Set up the directory of each run 
Turn path on and off using as each run is completed

```{r}
#path <- "../../Raw_data/Heron/FTP110/" # Complete!
#path <- "../../Raw_data/Heron/FTP126/" # Complete!
#path <- "../../Raw_data/Heron/FTP127/" # Complete!
#path <- "../../Raw_data/Heron/FTP132/" # Complete!

list.files(path) # Check before moving onwards that you have the right path established
```

## Check for primers

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern = "R1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "R2.fastq", full.names = TRUE))
```

```{r}
# Define Symbiodinium ITS2 primers
FWD <- "GTGAATTGCAGAACTCCGTG"
REV <- "CCTCCGCTTACTTATATGCTT"
```

```{r}
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = Biostrings::complement(dna), Reverse = Biostrings::reverse(dna),
        RevComp = Biostrings::reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
REV.orients
```

```{r}
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs[[1]]),
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs[[1]]),
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs[[1]]),
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs[[1]]))

# After running the primer search function above on the demultiplexed files we see that all the primers are still on the forwards and reverse reads. Use cutadapt to remove these.
```

## Cutadapt bash script

Bash script to remove primers in all orientations:

```{}
bash cutadapt_trim_primers_b.bash
```

This will launch the looped code below:

```{}
for i in *.R1.fastq.gz
do
  SAMPLE=$(echo ${i} | sed "s/.R1\.fastq\.gz//") 
  echo ${SAMPLE}.R1.fastq.gz ${SAMPLE}.R2.fastq.gz
cutadapt -b file:primers.fasta -B file:primers.fasta -n 2 -m 50 -o ${SAMPLE}.R1.CUT.fastq.gz -p ${SAMPLE}.R2.CUT.fastq.gz ${SAMPLE}.R1.fastq.gz ${SAMPLE}.R2.fastq.gz
done
```

```{r}
# List new directory of 'Cuts'. These are the files generated by the bash script above that have .CUT. in the filename.
path <- "../../Raw_data/Heron/FTP110/cuts/"
path <- "../../Raw_data/Heron/FTP126/cuts/"
path <- "../../Raw_data/Heron/FTP127/cuts/"
path <- "../../Raw_data/Heron/FTP132/cuts/"


list.files(path)
```

```{r}
# Pattern search again for R1.CUT and R2.CUT in filenames
cutFs <- sort(list.files(path, pattern = "R1.CUT.fastq", full.names = TRUE))
cutRs <- sort(list.files(path, pattern = "R2.CUT.fastq", full.names = TRUE))
```

```{r}
# Check how well cutadapt worked at removing primers

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = cutFs[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = cutRs[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = cutFs[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = cutRs[[1]]))

## All primers removed!
```

```{r}
# Two rounds of strsplit to remove head and tail of the file name and keep the sample name.
sample.names <- sapply(strsplit(basename(cutFs), ".R1"), `[`, 1)
sample.names <- sapply(strsplit(sample.names, "demuxtrim-"), `tail`, 1)
```

```{r}
# Check quality of sequences of Forward reads
plotQualityProfile(cutFs[1:3])
```

```{r}
# Check quality of sequences of Reverse reads
plotQualityProfile(cutRs[1:3])
```

```{r}
# Set up subdirectory for placing the filtered sequences
filtFs <- file.path(path, "filtered", paste0(sample.names, "_R1_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R2_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

```{r}
# Filter the cutFs and cutRs into the directories created above. maxN = 0 because dada2 cannot have sequences with ambiguous bases. maxEE = c(2, 4): relaxed the EE to 4 for the reverse reads which typically are of lower quality.

out <- filterAndTrim(cutFs, filtFs, cutRs, filtRs, maxN = 0, maxEE = c(2, 4), truncQ = 2, minLen = 50, rm.phix = TRUE, compress = TRUE, multithread = FALSE) # On Windows set multithread=FALSE

# Check how many reads make it through filtering
head(out) 
```

```{r}
# Create an error model for the forward reads
errF <- learnErrors(filtFs, multithread = FALSE)
```

```{r}
# Create an error model for the reverse reads
errR <- learnErrors(filtRs, multithread = FALSE)
```

```{r}
# Check that the error model fits the errors
plotErrors(errF, nominalQ = TRUE)
```

## Denoise and infer samples

```{r}
# Run the core dada2 method to remove noise and infer samples on forward reads
dadaFs <- dada(filtFs, err = errF, multithread = FALSE)
```

```{r}
# Run the core dada2 method to remove noise and infer samples on reverse reads
dadaRs <- dada(filtRs, err = errR, multithread = FALSE)
```

```{r}
# Merge the forward and reverse reads
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose = TRUE)
```

```{r}
# Turn the successfully merged sequences into a sequence table (i.e. an 'otu table' or 'asv table')
seqtab <- makeSequenceTable(mergers)
```

## Save point

```{r}
#saveRDS(seqtab, "FTP110.rds") # Complete!
#saveRDS(seqtab, "FTP126.rds") # Complete!
#saveRDS(seqtab, "FTP127.rds") # Complete!
#saveRDS(seqtab, "FTP132.rds") # Complete!
```

# 3. Merge runs

```{r}
### Join each run + culture data into a single sequence table
FTP110 <- readRDS("FTP110.rds")
FTP126 <- readRDS("FTP126.rds")
FTP127 <- readRDS("FTP127.rds")
FTP132 <- readRDS("FTP132.rds")
Cultures <- readRDS("../dada2_cultures/complete_seqtab.rds")
complete_seqtab <- mergeSequenceTables(FTP110, FTP126, FTP127, FTP132, Cultures)
saveRDS(complete_seqtab, "complete_seqtab.rds")
```

# 4. Remove chimeras after run merging

```{r}
seqtab.nochim <- removeBimeraDenovo(complete_seqtab, method = "consensus", multithread = FALSE)
saveRDS(seqtab.nochim, "seqtab.nochim.rds")

sum(seqtab.nochim) / sum(complete_seqtab) # = 0.9578121
```

# 5. Assign taxonomy:

## Assign taxonomy to reads

```{r}
seqtab.nochim <- readRDS("seqtab.nochim.rds")

tax_bs <- assignTaxonomy(seqtab.nochim, "../../Raw_data/Reference_databases/ITS2dbn1_Dinophy_Symbio.fasta", multithread = FALSE, outputBootstraps = TRUE, minBoot = 0)
saveRDS(tax_bs, "seqtab.nochim.tax_bs.rds")
```

## Inspect bootstrap values

```{r}
tax_bs <- readRDS("seqtab.nochim.tax_bs.rds")
taxa <- as.data.frame(tax_bs$tax, stringsAsFactors = FALSE)
taxa <- tibble::rownames_to_column(taxa, var = "seq")

boot <- as.data.frame(tax_bs$boot)
boot <- tibble::rownames_to_column(boot, var = "seq")

# Inspect boostrap values

taxfilt <- left_join(taxa, boot, by = "seq", suffix = c("tax", "boot"))

# 50% bs support for symbiodiniaceae at family looks ok
# Below this it falls off to other dinoflagellates in NCBI

taxfilt <- left_join(taxa, boot, by = "seq", suffix = c("tax", "boot")) %>%
  mutate(Genustax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Genustax)),
         Speciestax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Speciestax)),
         Familytax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Familytax))) %>%
  mutate(Phylumtax = case_when(Phylumboot < 50  & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                               TRUE ~ as.character(.$Phylumtax)),
         Classtax = case_when(Classboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Classtax)),
         Ordertax = case_when(Orderboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Ordertax)),
         Familytax = case_when(Familyboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                               TRUE ~ as.character(.$Familytax)),
         Genustax = case_when(Genusboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Genustax)),
         Speciestax = case_when(Speciesboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                                TRUE ~ as.character(.$Speciestax)))

rownames(taxfilt) <- taxfilt$seq
taxfilt$seq = NULL

saveRDS(taxfilt, "seqtab.nochim.tax.rds")
```

# 6. Hand-off to phyloseq

```{r}
seqtab.nochim <- readRDS("seqtab.nochim.rds")

tax <- readRDS("seqtab.nochim.tax.rds") %>%
  select(Kingdomtax, Phylumtax, Classtax, Ordertax, Familytax, Genustax, Speciestax)
taxtab <- tax_table(tax)
rownames(taxtab) <- rownames(tax)
colnames(taxtab) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

samptab <- read_csv("metadata.csv")
rownames(samptab) <- samptab$Sample_ID
samptab <- sample_data(samptab)

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), 
               sample_data(samptab),
               taxtab)

dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps

saveRDS(ps, "phyloseq.rds")
```